src/clj/puppetlabs/services/legacy_routes/legacy_routes_core.clj

```diff
@@ -38,6 +38,32 @@
   (fn [request]
     (ring-response/content-type (handler request) "text/plain")))
 
+(defn- add-query-param
+  "Given a query string, add the given key and value to it. If the given
+  query string is empty or nil then the resulting string will be the single
+  KEY=VALUE expression."
+  [query-str key value]
+  {:pre [(or (string? query-str) (nil? query-str))
+         (or (string? key) (keyword? key))
+         (string? value)]
+   :post [(string? %)]}
+  (let [param-pairs (if (str/blank? query-str)
+                      {}
+                      (ring-codec/form-decode query-str))]
+    (ring-codec/form-encode (assoc param-pairs key value))))
+
+(defn- add-source-permissions-param
+  "Add the source_permissions header to the request. This is needed since
+  this was considered a default in Puppet 3.x and is now turned off by default
+  in Puppet 4.x. See SERVER-684."
+  [handler]
+  (fn [request]
+    (let [{:keys [query-string]} request
+          updated-request (assoc request :query-string
+                            (add-query-param
```

Comment:
If I understand this correctly, this would mean that if "source_permissions=X" were already in the query string that we would replace that with "source_permissions=use".  That may be irrelevant for Puppet 3.x agents if they never include the "source_permissions" parameter.  But if a client could include this parameter and the value it includes is trustworthy - i.e., we don't have the case where a client says "source_permissions=X" but really expects the server to process the request as though it had specified "source_permissions=use" - then I would rather have the implementation only add this query param if it wasn't already present in the request.  That way, we'd just be overriding the default behavior when the client doesn't specify a value rather than potentially processing the request with a value that the client did not intend.  @nwolfe - what do you think?